// css-styled/dist/styled.js
/*
Copyright (c) 2019 Daybrush
name: css-styled
license: MIT
author: Daybrush
repository: git+https://github.com/daybrush/css-styled.git
version: 0.1.7
*/
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('@daybrush/utils')) :
  typeof define === 'function' && define.amd ? define(['@daybrush/utils'], factory) :
  (global = global || self, global.styled = factory(global.utils));
}(this, (function (utils) { 'use strict';

  function hash(str) {
    var hash = 5381,
        i    = str.length;

    while(i) {
      hash = (hash * 33) ^ str.charCodeAt(--i);
    }

    /* JavaScript does bitwise operations (like XOR, above) on 32-bit signed
     * integers. Since we want the results to be always positive, convert the
     * signed int to an unsigned by doing an unsigned bitshift. */
    return hash >>> 0;
  }

  var stringHash = hash;

  function getHash(str) {
    return stringHash(str).toString(36);
  }
  function getShadowRoot(parentElement) {
    if (parentElement.getRootNode) {
      var rootNode = parentElement.getRootNode();

      if (rootNode.nodeType === 11) {
        return rootNode;
      }
    }

    return;
  }
  function injectStyle(className, css, shadowRoot) {
    var style = document.createElement("style");
    style.setAttribute("type", "text/css");
    style.innerHTML = css.replace(/([^}{]*){/mg, function (all, selector) {
      return utils.splitComma(selector).map(function (subSelector) {
        if (subSelector.indexOf(":global") > -1) {
          return subSelector.replace(/\:global/g, "");
        } else if (subSelector.indexOf(":host") > -1) {
          return "" + subSelector.replace(/\:host/g, "." + className);
        }

        return "." + className + " " + subSelector;
      }).join(", ") + "{";
    });
    (shadowRoot || document.head || document.body).appendChild(style);
    return style;
  }

  function styled(css) {
    var injectClassName = "rCS" + getHash(css);
    var injectCount = 0;
    var injectElement;
    return {
      className: injectClassName,
      inject: function (el) {
        var shadowRoot = getShadowRoot(el);
        var firstMount = injectCount === 0;
        var styleElement;

        if (shadowRoot || firstMount) {
          styleElement = injectStyle(injectClassName, css, shadowRoot);
        }

        if (firstMount) {
          injectElement = styleElement;
        }

        if (!shadowRoot) {
          ++injectCount;
        }

        return {
          destroy: function () {
            if (shadowRoot) {
              el.removeChild(styleElement);
              styleElement = null;
            } else {
              if (injectCount > 0) {
                --injectCount;
              }

              if (injectCount === 0 && injectElement) {
                injectElement.parentNode.removeChild(injectElement);
                injectElement = null;
              }
            }
          }
        };
      }
    };
  }

  return styled;

})));
